File: modules/utilities/url/c/ios/IAUrl.m
//
//  IAUrl.m
//  ImpsEngineiOS
//
//  Created by Konstantin Merker on 17.08.15.
//  Copyright (c) 2015 Konstantin Merker, Paul Wallrabe und Martin Krautschick GbR (Imps Apps). All rights reserved.
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>

static NSString * const iOS11AppStoreUrlFormat = @"itms-apps://itunes.apple.com/us/app/id%s?action=write-review";
static NSString * const iOS7AppStoreUrlFormat = @"itms-apps://itunes.apple.com/app/id%s";
static NSString * const iOSAppStoreUrlFormat = @"itms-apps://itunes.apple.com/WebObjects/MZStore.woa/wa/viewContentsUserReviews?type=Purple+Software&id=%s";

void IAUrl_openUrlInWebviewWithNSString(NSString * urlString);


void IAUrl_openUrlInWebview(const char * url){
    NSString * urlString = [NSString stringWithUTF8String:url];
    IAUrl_openUrlInWebviewWithNSString(urlString);
}

void IAUrl_openReviewPageInAppstore(const char * iosAppStoreId){
    NSString * appStoreUrlFormat;
    if ([[UIDevice currentDevice].systemVersion floatValue] >= 11.0f){
        appStoreUrlFormat = iOS11AppStoreUrlFormat;
    }else if ([[UIDevice currentDevice].systemVersion floatValue] >= 7.0f) {
        appStoreUrlFormat = iOS7AppStoreUrlFormat;
    }else{
        appStoreUrlFormat = iOSAppStoreUrlFormat;
    }
    NSString * appStoreUrlString = [NSString stringWithFormat:appStoreUrlFormat, iosAppStoreId];
    IAUrl_openUrlInWebviewWithNSString(appStoreUrlString);
}

void IAUrl_openUrlInWebviewWithNSString(NSString * urlString){
    NSURL * url = [NSURL URLWithString:urlString];
    [[UIApplication sharedApplication] openURL: url];
}

File: modules/utilities/url/c/android/IAUrl.c
//
//  Url.c
//  ImpsEngineAndroid
//
//  Created by Konstantin Merker on Mar 25, 2016.
//  Copyright (c) 2015 Konstantin Merker, Paul Wallrabe und Martin Krautschick GbR (Imps Apps). All rights reserved.
//


#include "IALibrary.h"
#include "IAUrl.h"
#include "JNIEnv+HelperMethods.h"
#include "IAJni.h"

static jclass jUrlClass;

static jmethodID openUrlInWebview;
static jmethodID openReviewPageInAppstore;

static bool isCommenced;

void IAUrl_commenceJniIfNeeded(JNIEnv * env){
    if (isCommenced == false){
        isCommenced = true;
        jUrlClass = JNIEnv_createNativeClass(env, "com/impsapps/impsengine/utilities/url/Url");
        openUrlInWebview = (*env)->GetStaticMethodID(env, jUrlClass, "openUrlInWebview", "(Ljava/lang/String;)V");
        openReviewPageInAppstore = (*env)->GetStaticMethodID(env, jUrlClass, "openReviewPageInAppstore", "()V");
    }
}

void IAUrl_openUrlInWebview(const char * url){
    JNIEnv * env = IAJni_getJNIEnv();
    IAUrl_commenceJniIfNeeded(env);
    jstring jUrl = JNIEnv_createStringWithUTFCharsLocally(env, url);
    (*env)->CallStaticVoidMethod(env, jUrlClass, openUrlInWebview, jUrl);
    JNIEnv_destroyStringWithUTFCharsLocally(env, jUrl);
}

void IAUrl_openReviewPageInAppstore(const char * iosAppStoreId){
    JNIEnv * env = IAJni_getJNIEnv();
    IAUrl_commenceJniIfNeeded(env);
    (*env)->CallStaticVoidMethod(env, jUrlClass, openReviewPageInAppstore);
}
